src/ci/stage-build.py

```diff
@@ -147,6 +151,69 @@ def build_rustc_perf(self):
     def supports_bolt(self) -> bool:
         return True
 
+    def run_tests(self):
+        """
+        After `dist` is executed, we extract its archived components into a sysroot directory,
+        and then use that extracted rustc as a stage0 compiler.
+        Then we run a subset of tests using that compiler, to have a basic smoke test which checks
+        whether the optimization pipeline hasn't broken something.
+        """
+        dist_dir = self.build_root() / "build" / "dist"
+
+        def extract_dist_dir(name: str) -> Path:
+            target_dir = Path("/tmp/dist")
+            target_dir.mkdir(parents=True, exist_ok=True)
+            unpack_archive(dist_dir / f"{name}.tar.xz", target_dir=target_dir)
+            extracted_path = target_dir / name
+            assert extracted_path.is_dir()
+            return extracted_path
+
+        # Extract rustc archive to get the optimized sysroot
+        extracted_rust_dir = extract_dist_dir(f"rust-nightly-{PGO_HOST}")
+        sysroot = Path("/tmp/dist-sysroot")
+        cmd([str(extracted_rust_dir / "install.sh"), f"--destdir={sysroot}", "--prefix=/", "--disable-ldconfig"])
+
+        # Extract sources - we need to manually copy libstd sources to the extracted sysroot
+        extracted_src_dir = extract_dist_dir(f"rust-src-nightly")
+        shutil.copytree(
+            extracted_src_dir / "rust-src" / "lib" / "rustlib" / "src",
+            sysroot / "lib" / "rustlib" / "src"
+        )
+        rustc_path = sysroot / "bin" / "rustc"
+        assert rustc_path.is_file()
+
+        # The SHA is needed for remapping compiler paths, which we have to do for
+        # some (ui and mir-opt) tests to pass.
+        with open(extracted_rust_dir / "git-commit-hash") as f:
+            commit_sha = f.read().strip()
+
+        # Simulate a stage 0 compiler with the extracted optimized dist artifacts.
+        with open("config.toml", "w") as f:
+            f.write(f"""
+[build]
+rustc = "{rustc_path}"
+""")
+
+        cmd(args=[
+            sys.executable,
+            self.checkout_path() / "x.py",
+            "test",
+            "--stage", "0",
+            "tests/assembly",
+            "tests/codegen",
+            "tests/codegen-units",
+            "tests/incremental",
+            "tests/mir-opt",
+            "tests/pretty",
+            "tests/ui",
```

Comment:
Does rustdoc also have PGO applied? Should we test it too?

Also, I'd expect `run-make` and `run-pass-valgrind` to be in this list - were they hard to get working?

Reply:
`rustdoc` isn't currently optimized, however it still usess the optimized `rustc_driver` library, so there could be some regressions. It was failing locally, I'll test it again with the latest changes.

Reply:
`run-make` fails with this:
```
error[E0464]: multiple candidates for `rlib` dependency `std` found
  |
  = note: candidate #1: /rustc/rust/build/build-ci/sysroot/lib/rustlib/x86_64-unknown-linux-gnu/lib/libstd-edb03adabf0b22c8.rlib
  = note: candidate #2: /rustc/rust/build/x86_64-unknown-linux-gnu/stage0-sysroot/lib/rustlib/x86_64-unknown-linux-gnu/lib/libstd-3a66bdf19d6deae0.rlib
```
same with `rustdoc` and `rustdoc-json`.

I suppose that it's because of this:
```
command: "/rustc/rust/build/build-ci/sysroot/bin/rustdoc" "-L" "/rustc/rust/build/x86_64-unknown-linux-gnu/stage0-sysroot/lib/rustlib/x86_64-unknown-linux-gnu/lib" "-L" "/rustc/rust/build/x86_64-unknown-linux-gnu/test/rustdoc/impl-everywhere/auxiliary" "-o" "/rustc/rust/build/x86_64-unknown-linux-gnu/test/rustdoc/impl-everywhere" "--deny" "warnings" "/rustc/rust/tests/rustdoc/impl-everywhere.rs"
```
stage0 library root being passed to the tests.
`run-pass-valgrind` works, `rustdoc-ui` too, except for one test.